Include simple context in triggerer async callback

ramitkataria opened the pull request:
- ~~Added jinja template rendering for all kwargs being passed to deadline async callbacks~~
- Added a simple context dict in the kwargs
- Set `context` as a reserved field for kwargs in callback definition

Eventually, we should probably use the TaskSDK API in the triggerer to fetch the full context but this solution covers most use cases for now.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

ashb reviewed (changes requested):
Please find a way to do this does doesn't involve importing SDK inside DB models. We need to:

a) not add new uses of SDK inside Core (i.e. anything that runs in the scheduler or API servers etc)
b) We don't really want to render templates inside the scheduler, as that is dangerously close to running user code there.

ramitkataria commented on airflow-core/src/airflow/triggers/deadline.py:
```diff
@@ -49,26 +49,25 @@ async def run(self) -> AsyncIterator[TriggerEvent]:
-            context: dict = {}
-
-            result = await callback(**self.callback_kwargs, context=context)
-            log.info("Deadline callback completed with return value: %s", result)
```
The triggerer already emits an info log when the TriggerEvent is yielded so we don't need an extra log here

ramitkataria commented on airflow-core/tests/unit/models/test_deadline.py:
```diff
@@ -232,20 +255,6 @@ def test_handle_callback_event(self, dagrun, deadline_orm, session, event, none_
         else:
             assert deadline_orm.callback_state == DeadlineCallbackState.QUEUED
 
-    def test_handle_miss_creates_trigger(self, dagrun, deadline_orm, session):
```
This tests a subset of what another one above tests

ramitkataria committed:
Include simple context in triggerer async callback

- Added a simple context dict in the kwargs
- Set `context` as a reserved field for kwargs in callback definition

Eventually, we should probably use the TaskSDK API in the triggerer to
fetch the full context but this solution covers most use cases for now.

ferruzzi commented on task-sdk/src/airflow/sdk/definitions/deadline.py:
```diff
@@ -120,8 +120,10 @@ class Callback(ABC):
-    def __init__(self, callback_callable: Callable | str, kwargs: dict | None = None):
+    def __init__(self, callback_callable: Callable | str, kwargs: dict[str, Any] | None = None):
         self.path = self.get_callback_path(callback_callable)
+        if kwargs and "context" in kwargs:
+            raise ValueError("context is a reserved kwarg for this class")
```
I like this, good thought.

ferruzzi reviewed (approved):
Looks good for now.  It'll be nice to get full context later but the dagrun has lots of useful info in it.

o-nikolas commented on airflow-core/src/airflow/triggers/deadline.py:
```diff
@@ -49,26 +49,25 @@ async def run(self) -> AsyncIterator[TriggerEvent]:
         try:
-            callback = import_string(self.callback_path)
             yield TriggerEvent({PAYLOAD_STATUS_KEY: DeadlineCallbackState.RUNNING})
+            callback = import_string(self.callback_path)
```
Any reason or impact for this swap around the yield?

ramitkataria replied:
Nothing significant - I just did it so that if there was a failure while importing and someone was monitoring the state of the callback, it would go from queued to running and then failed instead of queued -> failed

ashb commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -202,10 +202,20 @@ def handle_miss(self, session: Session):
             callback_trigger = DeadlineCallbackTrigger(
                 callback_path=self.callback.path,
-                callback_kwargs=self.callback.kwargs,
+                callback_kwargs=(self.callback.kwargs or {}) | {"context": get_simple_context()},
```
Nit: let's take anything called context from the user if they specify it? 
```suggestion
                callback_kwargs={"context": get_simple_context()} | (self.callback.kwargs or {}),
```

ramitkataria replied:
I set context as a reserved kwarg so that shouldn't be a problem, otherwise they wouldn't get this provided context in the callback, potentially without even knowing it exists

ashb commented on airflow-core/src/airflow/triggers/deadline.py:
```diff
@@ -49,26 +49,25 @@ async def run(self) -> AsyncIterator[TriggerEvent]:
                     PAYLOAD_BODY_KEY: f"{message}: {traceback.format_exception(e)}",
                 }
             )
+            log.exception("%s: %s; kwargs: %s\n%s", message, self.callback_path, self.callback_kwargs, e)
```
I think this should be before the yield, as if this is anything like how the triggers for tasks are handled, once the deadline is in a "terminal" state the async task will get cancelled and there is a chance the final log never shows up?

ramitkataria replied:
Good point - fixed!

ashb commented on airflow-core/tests/unit/models/test_deadline.py:
```diff
@@ -160,16 +161,37 @@ def test_repr_without_callback_kwargs(self, deadline_orm, dagrun, session):
+
+        kwargs = trigger.kwargs["callback_kwargs"]
+        context = kwargs.pop("context")
+        assert kwargs == kwargs
```
I don't think you meant to assign this to `kwargs` here, because L190 is not testing anything anymore, because of course `kwargs == kwargs` :)

ramitkataria replied:
Thanks for catching that, also fixed

ashb reviewed (approved):
Few nits, but LGTM now, thanks for changing it

ramitkataria committed:
PR nits

ramitkataria committed:
Merge branch 'main' into ramitkataria/deadlines/context-in-callback

kaxil commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -202,10 +202,20 @@ def handle_miss(self, session: Session):
+            # TODO: Use the TaskAPI from within Triggerer to fetch full context instead of sending this context
+            #  from the scheduler
+            return {
+                "dag_run": DAGRunResponse.model_validate(self.dagrun).model_dump(mode="json"),
```
nit: 
```suggestion
                "dag_run": DAGRunResponse.model_validate(self.dagrun).model_dump_json(),
```

kaxil replied:
You don't need to change it though if PR goes green on CI :)

kaxil reviewed (approved):
Yeah, this is fine for now. We can improve this post 3.1
